src/openapi/authProvider.ts

```diff
@@ -0,0 +1,465 @@
+    tokenRefreshUrl?: string;
+    tokenRefreshHeaders?: Record<string, string> = {};
+    tokenRefreshCredentials: RequestCredentials = DEFAULT_TOKEN_REFRESH_CREDENTIALS;
+    tokenRefreshMethod: HTTPMethodType = DEFAULT_TOKEN_REFRESH_METHOD;
+    tokenRefreshPropertyNameToken = DEFAULT_TOKEN_REFRESH_PROPERTY_NAME_TOKEN;
+    tokenRefreshPropertyNameExpires = DEFAULT_TOKEN_REFRESH_PROPERTY_NAME_EXPIRES;
+    tokenRefreshMarginMs = DEFAULT_TOKEN_REFRESH_MARGIN_MS;
+    retryDelayMs = DEFAULT_RETRY_DELAY_MS;
+    maxRetryCount = DEFAULT_MAX_RETRY_COUNT;
+    state = 1;
+    retries = 0;
+    tokenRefreshTimerFireTime = 0;
+    tokenRefreshTimer: number | null = null;
+    lastTokenFetchTime = 0;
+    // Type of event that occurs when the token is refreshing.
+    EVENT_TOKEN_REFRESH = EVENT_TOKEN_REFRESH;
+    // Type of event that occurs when the token is received.
+    EVENT_TOKEN_RECEIVED = EVENT_TOKEN_RECEIVED;
+    // Type of event that occurs when the token refresh fails.
+    EVENT_TOKEN_REFRESH_FAILED = EVENT_TOKEN_REFRESH_FAILED;
+
+    constructor(options: Options) {
+        super();
+        if (!options?.token && !options?.tokenRefreshUrl) {
+            throw new Error('No token supplied and no way to get it');
+        }
+
+        if (options?.expiry) {
+            this.expiry = options.expiry;
+        }
+
+        this.token = addBearer(options.token || null);
+
+        // convert to absolute if the value is too small to be absolute
+        if (this.expiry < Date.UTC(2000, 0)) {
+            this.expiry = toAbsoluteTokenExpiry(this.expiry);
+        }
+
+        this.tokenRefreshUrl = options.tokenRefreshUrl;
+        this.tokenRefreshHeaders = options.tokenRefreshHeaders;
+
+        if (options.tokenRefreshCredentials) {
+            this.tokenRefreshCredentials = options.tokenRefreshCredentials;
+        }
+        if (options.tokenRefreshMethod) {
+            this.tokenRefreshMethod = options.tokenRefreshMethod;
+        }
+        if (options.tokenRefreshPropertyNameToken) {
+            this.tokenRefreshPropertyNameToken =
+                options.tokenRefreshPropertyNameToken;
+        }
+        if (options.tokenRefreshPropertyNameExpires) {
+            this.tokenRefreshPropertyNameExpires =
+                options.tokenRefreshPropertyNameExpires;
+        }
+
+        if (options.tokenRefreshMarginMs) {
+            this.tokenRefreshMarginMs = options.tokenRefreshMarginMs;
+        }
+        if (options.retryDelayMs) {
+            this.retryDelayMs = options.retryDelayMs;
+        }
+        if (options.maxRetryCount) {
+            this.maxRetryCount = options.maxRetryCount;
+        }
+
+        this.state = STATE_WAITING;
+        this.retries = 0;
+
+        this.createTimerForNextToken();
+    }
+
+    /**
+     * This internal method refreshes the token no matter what and should only
+     * be called if we know we are in the correct state to do it.
+     */
+    private refreshToken = () => {

```

Comment:
yes, the arrow is used since it's passed to setTimeout. I used constructor binding in one place since I noticed that IDE provides better params/doc hints for methods but in this case, it doesn't matter.